core/src/main/java/com/linecorp/armeria/internal/common/stream/PrependingPublisher.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public final class PrependingPublisher<T> implements Publisher<T> {
+
+    private final T first;
+    private final Publisher<? extends T> rest;
+
+    public PrependingPublisher(T first, Publisher<? extends T> rest) {
+        this.first = first;
+        this.rest = rest;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        final RestSubscriber restSubscriber = new RestSubscriber(subscriber);
+        rest.subscribe(restSubscriber);
+    }
+
+    final class RestSubscriber implements Subscriber<T>, Subscription {
+
+        private Subscriber<? super T> downstream;
+        private final AtomicLong demand = new AtomicLong();
```

Comment:
Sorry for the late response 😄 To clarify, (maybe you know this already, but just to be sure we're on the same page) `demand` is used to satisfy rule [3.3](https://github.com/reactive-streams/reactive-streams-jvm#3.3), to avoid overflow in case the subscriber calls `request` in `onNext`. If `demand.getAndAdd(n) > 0`, i.e we already have an ongoing request, we should stop. It doesn't really concern `first`.

Reply:
Ah. I got it. 👍 How about using `AtomicLongFieldUpdater` with a `volatile` variable to reduce memory footprint?